File: updated_code/code/python/020-valid-parentheses.py
class ValidParentheses:
    """
    This class determines if the input string is valid.
    """
    def __init__(self):
        """
        Constructor for the ValidParentheses class.
        """
        pass

    def is_valid(self, s: str) -> bool:
        """
        Determine if the input string is valid.
        
        Parameters:
        s (str): The input string containing parentheses.
        
        Returns:
        bool: True if the string is valid, False otherwise.
        """
        if len(s) % 2 != 0:
            return False
        
        stack = []
        mapping = {")": "(", "}": "{", "]": "["}

        for char in s:
            if char in mapping.values():
                stack.append(char)
            elif char in mapping:
                if not stack or mapping[char] != stack.pop():
                    return False
        
        return not stack

File: updated_code/code/java/020-valid-parentheses.java
package com.example;

import java.util.Stack;

/**
 * Given a string `s` containing just the characters `'('`, `')'`, `'{'`, `'}'`,
 * `'['` and `']'`, determine if the input string is valid.
 *
 * An input string is valid if:
 * 1. Open brackets must be closed by the same type of brackets.
 * 2. Open brackets must be closed in the correct order.
 * 3. Every close bracket has a corresponding open bracket of the same type.
 */
public class Solution {
    /**
     * Determines if the input string is valid.
     *
     * @param s the input string
     * @return true if the input string is valid, false otherwise
     */
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (final char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c == '(' ? ')' : c == '{' ? '}' : ']');
            } else if (stack.isEmpty() || stack.pop() != c) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}